test/hotspot/jtreg/compiler/c2/irTests/TestVectorizeURShiftSubword.java

```diff
@@ -0,0 +1,162 @@
+/*
+ * Copyright (c) 2022, Arm Limited. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package compiler.c2.irTests;
+
+import compiler.lib.ir_framework.*;
+import java.util.Random;
+import jdk.test.lib.Asserts;
+import jdk.test.lib.Utils;
+
+/*
+ * @test
+ * @bug 8283307
+ * @key randomness
+ * @summary Auto-vectorization enhancement for unsigned shift right on signed subword types
+ * @requires os.arch=="amd64" | os.arch=="x86_64" | os.arch=="aarch64"
```

Comment:
This IR test for vectorizable check looks good on AArch64. But AFAIK, some operations cannot be vectorized on old x86 CPUs with AVX=1. Could you add something like `(os.simpleArch == "x64" & vm.cpu.features ~= ".*avx2.*")` to check the CPU feature?

Reply:
@merykitty @sviswa7 Could you help confirm if byte/short shift operations are vectorizable with all AVX versions of x86?

Reply:
@pfustc They are available either directly or through a series of instructions for majority of the architectures. All implemented in x86.ad. The match_rule_supported and match_rule_supported_vector will appropriately return true/false.

Reply:
RShiftVB would fail with `UseSSE < 4`.
But is there a x86 machine running jtreg tests with `UseSSE < 4`?

For `RShiftV{B,S}`, neither `match_rule_supported ` nor `match_rule_supported_vector ` would return false on x86 with AVX=1.
So the test seems fine on x86, right? @sviswa7 @pfustc

Reply:
> But is there a x86 machine running jtreg tests with `UseSSE < 4`?
> 
To clarify: is there a x86 cpu which only supports `UseeSSE < 4` ISA running jtreg tests?

Reply:
We do run vectorization (superword, vectorization, VectorAPI) tests in our testing with:
`-XX:UseAVX=0 -XX:UseSSE=3` and `-XX:UseAVX=0 -XX:UseSSE=2`

Reply:
> We do run vectorization (superword, vectorization, VectorAPI) tests in our testing with: `-XX:UseAVX=0 -XX:UseSSE=3` and `-XX:UseAVX=0 -XX:UseSSE=2`


`compiler/c2/irTests/TestVectorizeURShiftSubword.java` still passed with `-XX:UseAVX=0 -XX:UseSSE=3` and `-XX:UseAVX=0 -XX:UseSSE=2`.

Maybe the IR test framework wouldn't accept VM args like `-XX:UseAVX=? -XX:UseSSE=?`?

Reply:
Two questions @fg1417  :
1. Is `RShiftVB` generated with `-XX:UseAVX=0 -XX:UseSSE=2` on x86?
2. If the answer is no for question 1, why `TestVectorizeURShiftSubword.java` get passed?
Thanks.

Reply:
> Two questions @fg1417 :
> 
> 1. Is `RShiftVB` generated with `-XX:UseAVX=0 -XX:UseSSE=2` on x86?
> 2. If the answer is no for question 1, why `TestVectorizeURShiftSubword.java` get passed?
>    Thanks.

@DamonFool , no `RShiftB` nodes are generated with `-XX:UseAVX=0 -XX:UseSSE=2` on x86. The test passed because when adding VM options `-XX:UseAVX=0 -XX:UseSSE=2` externally, the IR framework won't do the IR check and just run the java code, I verified and made the guess. But I'm not clear whether all external VM options would stop the IR framework to do the IR check.

Reply:
> 
> @DamonFool , no `RShiftB` nodes are generated with `-XX:UseAVX=0 -XX:UseSSE=2` on x86. The test passed because when adding VM options `-XX:UseAVX=0 -XX:UseSSE=2` externally, the IR framework won't do the IR check and just run the java code, I verified and made the guess. But I'm not clear whether all external VM options would stop the IR framework to do the IR check.

Thanks for the verification.
Then, I think the test is fine.

What do you think? @vnkozlov

Reply:
As @fg1417 pointed, IR framework will not do any IR verification testing (it does not run test) when some flags are passed to test and not whitelisted (`UseAVX` and `UseSSE` are not on list):
https://github.com/openjdk/jdk/blob/master/test/hotspot/jtreg/compiler/lib/ir_framework/TestFramework.java#L105

I hacked VM to set `UseSSE=2 UseAVX=0` and test failed as expected because it did not find `RShiftVB` as you said:
```
One or more @IR rules failed:
Failed IR Rules (2) of Methods (2)
```
With `UseSSE=4 UseAVX=0` combination test passed. Our testing systems don't have anything less than SSE4 CPUs. That is why it passed our testing. I assume Github actions don't have such old systems too. 

**I don't see these changes passed pre-submit GitHub testing. Make sure they passed.**

Currently there are changes in review #8999 which allow to specify CPU features as condition for testing method.

I suggest to leave test as it is (if pre-submit GitHub testing passed). If we hit any issues later we will use #8999 API to fix it.

Reply:
> I don't see these changes passed pre-submit GitHub testing. Make sure they passed.

I don't know why the pre-submit GitHub testing didn't run in this PR.
So I create a draft PR for this path to run the pre-submit testing: https://github.com/openjdk/jdk/pull/9026 .

Reply:
> As @fg1417 pointed, IR framework will not do any IR verification testing (it does not run test) when some flags are passed to test and not whitelisted (`UseAVX` and `UseSSE` are not on list):
> https://github.com/openjdk/jdk/blob/master/test/hotspot/jtreg/compiler/lib/ir_framework/TestFramework.java#L105

Ah, I see.
Thanks @vnkozlov for your explanation.

Reply:
> > I don't see these changes passed pre-submit GitHub testing. Make sure they passed.
> 
> I don't know why the pre-submit GitHub testing didn't run in this PR. So I create a draft PR for this patch to run the pre-submit testing: #9026 .

The pre-submit testing finished without regression. @vnkozlov @fg1417 
Thanks.

Reply:
Thanks for your help on GHA testing @DamonFool .